### PR TITLE
Core/PacketsIO: Implemented or updated most of party related packets.

### PR DESCRIPTION
**Pull request by et65**
Packets updated or implemented :
- SMSG_INSTANCE_INFO : updated
- CMSG_SAVE_CUF_PROFILES : updated
- SMSG_LOAD_CUF_PROFILES : updated
- SMSG_PARTY_COMMAND_RESULT : updated
- CMSG_PARTY_INVITE : updated
- SMSG_PARTY_INVITE : updated
- CMSG_PARTY_INVITE_RESPONSE : updated
- CMSG_PARTY_UNINVITE : updated
- SMSG_GROUP_UNINVITE : updated
- CMSG_LEAVE_GROUP : updated
- SMSG_GROUP_DECLINE : updated
- SMSG_GROUP_DESTROYED : updated
- CMSG_MINIMAP_PING : updated
- SMSG_MINIMAP_PING : updated
- CMSG_CONVERT_RAID : updated
- CMSG_SET_EVERYONE_IS_ASSISTANT
- CMSG_DO_READY_CHECK : updated
- CMSG_READY_CHECK_RESPONSE : updated
- SMSG_READY_CHECK_COMPLETED : updated
- SMSG_READY_CHECK_RESPONSE : updated
- SMSG_READY_CHECK_STARTED : updated
- CMSG_REQUEST_PARTY_JOIN_UPDATES : implemented (source : sniffs)
- CMSG_REQUEST_PARTY_MEMBER_STATE : updated
- SMSG_PARTY_MEMBER_STATE : updated
- SMSG_PARTY_UPDATE : updated
- CMSG_REQUEST_RAID_INFO : updated
- CMSG_INITIATE_ROLE_POLL : updated
- SMSG_ROLE_POLL_INFORM : updated
- CMSG_SET_ROLE : updated
- SMSG_ROLE_CHANGED_INFORM : updated
- CMSG_CHANGE_SUB_GROUP : updated
- CMSG_SWAP_SUB_GROUPS : implemented
- CMSG_SET_ASSISTANT_LEADER : updated
- CMSG_SET_PARTY_LEADER : updated
- SMSG_GROUP_NEW_LEADER : updated
- CMSG_CLEAR_RAID_MARKER : implemented
- SMSG_RAID_MARKERS_CHANGED : implemented
- CMSG_UPDATE_RAID_TARGET : updated
- SMSG_SEND_RAID_TARGET_UPDATE_ALL : updated
- SMSG_SEND_RAID_TARGET_UPDATE_SINGLE : updated
- CMSG_OPT_OUT_OF_LOOT : updated
- CMSG_SET_LOOT_METHOD : updated

About group update flags:
- Not sure they are use for now.
- Pets now have their group update flags.
- Pets'power is no send to client anymore.
- Changes about them are inspired from SMSG_PARTY_MEMBER_STATS parsing of WowPacketParser, but it seems this packet is not use anymore.

CHAT_MSG_RAID has been fixed.

About Ready check:
- Correctly implemented this function.
- An update function has been added to Group class, and to GroupMgr class in order to manage the ready check expiration (when 35 seconds were gone, players who have not answered must be reported as AFK)

About Raid markers:
- Old spell effect SPELL_EFFECT_SUMMON_OBJECT_SLOT3 has been renamed to SPELL_EFFECT_CHANGE_RAID_MARKER and implemented. I'm sure about that because raid markers spells are the only spells that have this effect type.

Source: WowPacketParser, and sniffs from official.

### CODE REVIEW

File: src/server/game/Groups/Group.cpp [hunk 1]
```diff
@@ -1767,6 +1765,49 @@ void Group::ChangeMembersGroup(ObjectGuid guid, uint8 group)
...
+    uint8 tmp = slots[0]->group;
+    slots[0]->group = slots[1]->group;
+    slots[1]->group = tmp;
+
+    for (uint8 i = 0; i < 2; i++)
+    {
+        // Preserve new sub group in database for non-raid groups
+        if (!isBGGroup() && !isBFGroup())
+        {
+            PreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_UPD_GROUP_MEMBER_SUBGROUP);
```
Comment (Shauren) [inline src/server/game/Groups/Group.cpp]: this definitely should be in a transaction - you are executing 2 queries that only make sense together

**Comment by et65** [comment]
@Shauren: I've hesitated about that, I will edit, thanks you :)

File: src/server/game/Groups/Group.cpp (hunk 1 above)
Comment (et65) [inline src/server/game/Groups/Group.cpp]: I've hesitated about that, I will edit, thanks you :)

File: src/server/game/Server/Packets/MiscPackets.cpp [hunk 2]
```diff
@@ -451,3 +451,71 @@ WorldPacket const* WorldPackets::Misc::Dismount::Write()
...
     return &_worldPacket;
 }
+
+void WorldPackets::Misc::SaveCUFProfiles::Read()
+{
+    uint8 strLen;
+
+    _worldPacket >> Count;
+
+    for (uint8 i = 0; i < Count; i++)
```
Comment (Shauren) [inline src/server/game/Server/Packets/MiscPackets.cpp]: exploitable crash point - verify that Count < MAX_CUF_PROFILES

Comment (et65) [inline src/server/game/Server/Packets/MiscPackets.cpp]: It's already checked in WorldSession::HandleSaveCUFProfiles, but I can add a check if you want...

File: src/server/game/Server/Packets/MiscPackets.cpp [hunk 3]
```diff
@@ -451,3 +451,74 @@ WorldPacket const* WorldPackets::Misc::Dismount::Write()
...
+    uint8 strLen;
+
+    _worldPacket >> Count;
+
+    if (Count > MAX_CUF_PROFILES)
+        return;
+
+    for (uint8 i = 0; i < Count; i++)
+    {
+        CUFProfiles[i] = new CUFProfile;
```
Comment (Shauren) [inline src/server/game/Server/Packets/MiscPackets.cpp]: leaks memory if packet content is manipulated to be too short resulting in ByteBufferException being thrown - std::unique_ptr would be a good thing to use here

File: src/server/game/Server/Packets/PartyPackets.cpp [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+void WorldPackets::Party::PartyInviteResponse::Read()
+{
+    _worldPacket >> PartyIndex;
+
+    _worldPacket.ResetBitPos();
+
+    Accept = _worldPacket.ReadBit();
+    HasRolesDesired = _worldPacket.ReadBit();
+
+    if (HasRolesDesired)
```
Comment (Shauren) [inline src/server/game/Server/Packets/PartyPackets.cpp]: looks like a job for Optional

File: src/server/game/Server/Protocol/Opcodes.h [hunk 5]
```diff
@@ -571,7 +571,7 @@ enum OpcodeClient : uint32
     CMSG_REQUEST_HONOR_STATS                          = 0x1207,
     CMSG_REQUEST_LFG_LIST_BLACKLIST                   = 0x0A02,
     CMSG_REQUEST_PARTY_JOIN_UPDATES                   = 0x0620,
-    CMSG_REQUEST_PARTY_MEMBER_STATS                   = 0x0EB8,
+    CMSG_REQUEST_PARTY_MEMBER_STATE                   = 0x0EB8,
```
Comment (Shauren) [inline src/server/game/Server/Protocol/Opcodes.h]: **NO** renaming opcodes - the names currently present here are final (except unknowns)

File: src/server/game/Server/Packets/PartyPackets.h [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+        {
+        public:
+            GroupDecline(std::string const& name) : ServerPacket(SMSG_GROUP_DECLINE, 2 + name.size()), Name(name) { }
+
+            WorldPacket const* Write() override;
+
+            std::string Name;
+        };
+
+        class RequestPartyMemberState final : public ClientPacket
```
Comment (Shauren) [inline src/server/game/Server/Packets/PartyPackets.h]: build 18179 - 1015FA430 UserClientRequestPartyMemberStats::UserClientRequestPartyMemberStats

File: src/server/game/Server/Packets/MiscPackets.cpp (hunk 3 above)
Comment (et65) [inline src/server/game/Server/Packets/MiscPackets.cpp]: I've never used std::unique_ptr, I don't really understand when memory leaks can appear. Do you have some examples?

File: src/server/game/Entities/Pet/Pet.cpp [hunk 7]
```diff
@@ -1938,6 +1937,32 @@ void Pet::SetDisplayId(uint32 modelId)
...
+}
+
+void Pet::SetGroupUpdateFlag(uint32 flag)
+{
+    if (!GetOwner())
+        return;
+
+    Player* playerOwner = GetOwner()->ToPlayer();
+
+    if (playerOwner && playerOwner->GetGroup())
```
Comment (joschiwald) [inline src/server/game/Entities/Pet/Pet.cpp]: owner is always a valid player

Comment (et65) [inline src/server/game/Entities/Pet/Pet.cpp]: Owner can't be a creature ?

File: src/server/game/Spells/SpellEffects.cpp [hunk 8]
```diff
@@ -4648,6 +4648,27 @@ void Spell::EffectPullTowards(SpellEffIndex /*effIndex*/)
...
+
+    Player* player = m_caster->ToPlayer();
+    if (!player || !m_targets.HasDst())
+        return;
+
+    Group* group = player->GetGroup();
+    if (!group || (group->isRaidGroup() && !group->IsLeader(player->GetGUID()) && !group->IsAssistant(player->GetGUID())))
+        return;
+
+    uint8 markerId = effectInfo->BasePoints;
```
Comment (joschiwald) [inline src/server/game/Spells/SpellEffects.cpp]: it's Spell::damage

File: src/server/game/Groups/Group.h [hunk 9]
```diff
@@ -160,6 +173,20 @@ struct InstanceGroupBind
     InstanceGroupBind() : save(NULL), perm(false) { }
 };
 
+struct RaidMarker
+{
+    Position Pos;
+    uint32 MapId;
```
Comment (joschiwald) [inline src/server/game/Groups/Group.h]: you could use a WorldLocation (Position + mapid)

File: src/server/game/Entities/Pet/Pet.cpp (hunk 7 above)
Comment (joschiwald) [inline src/server/game/Entities/Pet/Pet.cpp]: https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/game/Entities/Pet/Pet.cpp#L45
https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/game/Entities/Pet/Pet.h#L145

Comment (et65) [inline src/server/game/Entities/Pet/Pet.cpp]: Okay, I made a confusion with Guardians I think, thanks you

**Comment by DDuarte** [comment]
@et65: come to irc (#trinity @ irc.rizon.net) when you got 5 minutes, pretty please

**Comment by et65** [comment]
@DDuarte : I'm on IRC ;)

**Comment by DDuarte** [comment]
@Shauren: I've changed to std::unique_ptr in 57bed9fb8daf3d125cfb9a7952593b23371d9d34, could you check that the usage is correct?

P.S I've also added make_unique, sorry about that.

**Comment by Frytiks** [comment]
Very Nice work :) maybe next your work will be battlegrounds packets ? :D

File: src/server/game/Server/Packets/ChannelPackets.h [hunk 10]
```diff
@@ -145,6 +145,56 @@ namespace WorldPackets
...
+            ObjectGuid UpdatedUserGUID;
+
+            ChannelFlags ChannelFlags = CHANNEL_FLAG_NONE;
+            ChannelMemberFlags UserFlags = MEMBER_FLAG_NONE;
+
+            int32 ChannelID = 0;
+
+            std::string ChannelName;
+        };
+
```
Comment (Ladel) [inline src/server/game/Server/Packets/ChannelPackets.h]: @DDuarte double code

**Comment by et65** [comment]
@DDuarte : Thanks for your help, I known understand better the utility and how std::unique_ptr work :).

@DJScias : It's a voluntary mistake, it's because class Packet already has a method named Initialize, so we can't use this name. I know it's crapy but it's also done in ChatPackets (Chat::Initalize). Rename this method make build failed :(

**Comment by DDuarte** [comment]
Oops.

**Comment by DJScias** [comment]
@DDuarte & @et65 Sorry, my mistake then! It just triggered me :(

**Comment by et65** [comment]
@DJScias : No I think you're right, I made a confusion, Initialize is a method of WorldPacket and not Packet so it could be good to use Initialize in Packet childs. Note: so there is also a mistake in ChatPackets (https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/game/Server/Packets/ChatPackets.cpp#L95)

**Comment by et65** [comment]
Rebased into one commit. I think everything is ok right now :)

**Pull request merged**